### PR TITLE
add AudioContextOptions to AudioContext

### PR DESCRIPTION
**Pull request by Jerboas86**
The user can now specify:
- the number of output channels  
- the sample rate
- the output latency
when building its `AudioContext`.

If the intented options are not met by the audio hardware capabilities, the `AudioContext` is built with `default_config` options, and a warning is printed to the user.

If the user options are missing, the `AudioContext` is built with `primary_config` options. If the `primary_config` options are not met by the audio harware capabilities, the `AudioContext` is built with `default_config` options, and a warning is printed to the user.

This PR modify `DestinationNode`:
- Adds `max_channels_count` method
- Assert `ChannelCountConstraints` on `DestinationNode`

This PR adds merger.rs example to demonstrate new capabilities.

Clippy passes on `context.rs` and `io.rs`

related to #35
related to #28

### CODE REVIEW

File: src/io.rs [hunk 1]
```diff
@@ -89,27 +117,83 @@ pub(crate) fn build_output(
...
+    primary_config.buffer_size = cpal::BufferSize::Fixed(buffer_size);
+
+    let mut config: StreamConfig = supported_config.clone().into();
+
+    match options {
+        Some(opts) => {
+            config.sample_rate = CpalSampleRate(opts.sample_rate.unwrap_or(config.sample_rate.0));
+            config.channels = opts.channels.unwrap_or(config.channels);
+
+            match opts.latency_hint {
```
Comment (orottier) [inline src/io.rs:130]: Can you move this block into a helper function, makes it a bit more readable and then we can re-use if for the InputBuilder

File: src/io.rs [hunk 2]
```diff
@@ -134,7 +219,8 @@ pub(crate) fn build_output(
     (stream, config, sender)
 }
 
-pub(crate) fn build_input() -> (Stream, StreamConfig, Receiver<AudioBuffer>) {
+/// Builds the input
+pub fn build_input() -> (Stream, StreamConfig, Receiver<AudioBuffer>) {
```
Comment (orottier) [inline src/io.rs:223]: You should port the latency hint to the input builder too (can be done in another PR if it's too much work). They should be symmetrical in the end

File: src/context.rs [hunk 3]
```diff
@@ -488,11 +590,15 @@ impl BaseAudioContext {
             outputs: node.number_of_outputs() as usize,
             channel_config: node.channel_config_cloned(),
         };
-        self.inner.render_channel.send(message).unwrap();
+        self.inner
+            .render_channel
+            .send_timeout(message, Duration::from_millis(10))
```
Comment (orottier) [inline src/context.rs:595]: Why opt for timeout? If the requested latency `> 10millis`, this will fail for no good reason

File: src/context.rs [hunk 4]
```diff
@@ -575,21 +691,27 @@ impl OfflineAudioContext {
...
             length,
-            render,
+            renderer,
         }
     }
 
+    /// `OfflineAudioContext` doesn't start rendering automatically
+    /// You need to call this function to start the audio rendering
     pub fn start_rendering(&mut self) -> AudioBuffer {
         // make buffer_size always a multiple of BUFFER_SIZE, so we can still render piecewise with
```
Comment (orottier) [inline src/context.rs:701]: I think actually this is no longer needed, now we can handle varying buffer lengths.
We can remove this code in another PR

File: src/context.rs [hunk 5]
```diff
@@ -575,21 +691,27 @@ impl OfflineAudioContext {
...
-        let _split = buf.split_off(self.length as u32);
+        let mut buf = self.renderer.render_audiobuffer(buffer_size);
+        let _split = buf.split_off(self.length);
         buf
     }
 
-    pub fn length(&self) -> usize {
+    /// get the length of rendering audio buffer
+    #[must_use]
+    pub const fn length(&self) -> usize {
```
Comment (orottier) [inline src/context.rs:714]: no const please

**Review by orottier — CHANGES_REQUESTED** [review]
Interesting work, thanks!

File: src/io.rs [hunk 6]
```diff
@@ -57,84 +85,356 @@ fn spawn_input_stream(
...
-        receiver,
-        frames_played_clone,
-    );
+    /// modify the buffer size config, following user options
+    ///
+    /// # Argument
+    ///
+    /// * `options` - options contains latency hint information
+    ///
+    /// # Warning
```
Comment (Jerboas86) [inline src/io.rs:209]: As mentioned here the correctness of the API is not enforce by the type system. I could make it more strict but at the cost of more complexity. What do you think ?

**Review by Jerboas86 — COMMENTED** [review]
> You should port the latency hint to the input builder too (can be done in another PR if it's too much work)

I didn't port the refactoring to the input stream. leave it as "good first issue" or for another PR, provide that you agree on the refactoring :)

**Comment by Jerboas86** [comment]
> I didn't port the refactoring to the input stream. i leave it as "good first issue" or for another PR
 
is it Ok with you, if i open a good first issue about this topic?

**Comment by orottier** [comment]
Go ahead!

**Pull request merged**